contracts/crowdsale/XctCrowdsale.sol

```diff
@@ -0,0 +1,197 @@
+   */
+  uint256 public startTimeStageOne;
+
+  uint256 public startTimeStageTwo;
+
+  /**
+   * End timestamp to end the crowdsale.
+   */
+  uint256 public endTime;
+
+  /**
+   * Maximum amount of wei to spend during the first stage of the crowdsale.
+   */
+  uint256 public personalWeiCap;
+
+  /**
+   * Crowdsale cap in wei.
+   */
+  uint256 public crowdSaleWeiCap;
+  /**
+   * Raised amount in wei.
+   */
+  uint256 public weiRaised;
+
+  /**
+   * Address where funds are collected.
+   */
+  address public wallet;
+
+  /**
+   * How many token units a buyer gets per wei.
+   */
+  uint256 public rate;
+
+  /**
+   * @dev An event which is triggered when tokens are bought.
+   * @param _from The address sending tokens.
+   * @param _to The address recieving tokens.
+   * @param _weiAmount Purchase amount in wei.
+   * @param _tokenAmount The amount of purchased tokens.
+   */
+  event TokenPurchase(address indexed _from, address indexed _to, uint256 _weiAmount,
+                      uint256 _tokenAmount);
+
+  /**
+   * @dev Contract constructor.
+   */
+  function XctCrowdsale(uint256 _startTimeStageOne,
+                        uint256 _personalWeiCap,
+                        uint256 _crowdSaleWeiCap,
+                        uint256 _rate,
+                        address _walletAddress,
+                        address _tokenAddress)
+    public
+  {
+    require(_startTimeStageOne >= now);
+    require(_crowdSaleWeiCap > 0);
+    require(_personalWeiCap > 0);
+    require(_personalWeiCap < _crowdSaleWeiCap);
+    require(_rate > 0);
+    require(_walletAddress != address(0));
+    require(_tokenAddress != address(0));
+    require(_tokenAddress != _walletAddress);
+
+    token = Xct(_tokenAddress);
+    require(token.totalSupply() >= _crowdSaleWeiCap.mul(_rate));
+
+    startTimeStageOne = _startTimeStageOne;
+    startTimeStageTwo = _startTimeStageOne + 1 days;
+    endTime = startTimeStageTwo + 7 days;
+    personalWeiCap = _personalWeiCap;
+    crowdSaleWeiCap = _crowdSaleWeiCap;
+    rate = _rate;
+    wallet = _walletAddress;
+  }
+
+  /**
+   * @dev Fallback function can be used to buy tokens.
+   */
+  function()
+    external
+    payable
+  {
+    buyTokens(msg.sender);
+  }
+
+  /**
+   * @dev Low level token purchase function.
+   * @param beneficiary Address which is buying tokens.
+   */
+  function buyTokens(address beneficiary)
+    public
+    payable
+  {
+    require(beneficiary != address(0));
+    require(validPurchase(msg.value));
+
+    uint256 weiAmount = msg.value;
+
+    // calculate token amount to be created
+    uint256 tokens = getTokenAmount(weiAmount);
+
+    weiRaised = weiRaised.add(weiAmount);
+
+    if (!token.transferFrom(token.owner(), beneficiary, tokens)) {
+      revert();
+    }
+    TokenPurchase(msg.sender, beneficiary, weiAmount, tokens);
+
+    forwardFunds();
+  }
+
+  /**
+   * @return true if crowdsale event has ended
+   */
+  function hasEnded()
+    public
```

Comment:
This can probably be changed to external.

Reply:
https://github.com/0xcert/ethereum/pull/5/commits/b6633a9d33b1453bc2c3029280331c8ff0cb0488